packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -222,6 +275,116 @@ export class SchemaConverter implements SimpleDbSchemaConverter {
+    txn: SimpleDbTransaction,
+    fromVersion: number,
+    toVersion: number
+  ): PersistencePromise<void> {
+    let p = PersistencePromise.resolve();
+
+    /*
+     * Adding a new migration? READ THIS FIRST!
+     *
+     * Be aware that the SDK version may be downgraded then re-upgraded. This means that running
+     * your new migration must not prevent older versions of the SDK from functioning.
+     * Additionally, your migration must be able to run multiple times. In practice, this means a
+     * few things:
+     *  * Do not delete stores, change their key paths, or delete fields. Older versions may be
+     *    reading and writing them.
+     *  * Guard schema additions. Check if stores exist before adding them.
+     *  * Data migrations should *probably* always run. Older versions of the SDK will not have
+     *    maintained invariants from later versions, so migrations that update values cannot
+     *    assume that existing values have been properly maintained. Calculate them again, if
+     *    applicable.
+     */
+
+    /*
+     * As above, the migrations should be in the form of
+     * if (fromVersion < $VERSION && toVersion >= $VERSION) {
+     *   p = p.next(() => doMigrationSteps());
+     * }
+     */
+    if (fromVersion < 9 && toVersion >= 9) {
+      p = p.next(() => doTestMigration(txn));
+    }
+
+    return p.next(() => writeManualSchemaVersion(txn, toVersion));
+  }
+}
+
+function addManualSchemaVersion(
+  txn: SimpleDbTransaction,
+  db: IDBDatabase
+): PersistencePromise<void> {
+  ifStoresDontExist(db, [DB_VERSION_GLOBAL_STORE], () =>
+    db.createObjectStore(DB_VERSION_GLOBAL_STORE)
+  );
+  // Technically, this can't be a re-upgrade, since this is the last migration from IndexedDB's
+  // schema versioning. However, in order to follow the model of idempotent migrations, we would
+  // want to rewrite the manual schema version back to its initial value. This would allow future
+  // migrations the chance to run again.
+  return writeManualSchemaVersion(txn, FIRST_MANUAL_SCHEMA_VERSION);
+}
+
+function readManualSchemaVersion(db: IDBDatabase): PersistencePromise<number> {
+  const txn = new SimpleDbTransaction(
+    db.transaction(DB_VERSION_GLOBAL_STORE, 'readonly')
+  );
+  const store = txn.store<DbVersionGlobalKeyType, DbVersionGlobal>(
+    DB_VERSION_GLOBAL_STORE
+  );
+  return store.get(DB_VERSION_GLOBAL_KEY).next(maybeVersion => {
+    assert(maybeVersion !== null, 'Missing manual version');
+    return maybeVersion!;
+  });
+}
+
+function writeManualSchemaVersion(
+  txn: SimpleDbTransaction,
+  version: number
+): PersistencePromise<void> {
+  const store = txn.store<DbVersionGlobalKeyType, DbVersionGlobal>(
+    DB_VERSION_GLOBAL_STORE
+  );
+  return store.put(DB_VERSION_GLOBAL_KEY, version);
+}
+
+function doTestMigration(txn: SimpleDbTransaction): PersistencePromise<void> {
+  // Dummy migration to test a version number not based on indexeddb's versioning scheme.
```

Comment:
Suggested rewording to be a bit more explicit:

// Dummy migration that does nothing. It only exists for tests to exercise our "manual" version number mechanism and the ability to downgrade / re-upgrade successfully.